Introduce IcebergFileWriter and collect file stats directly from ORC writers

lxynov opened the pull request:
Umbrella issue: #1324 

### Commit `Fix predicating on REAL type in Iceberg Connector`:
Previously there was no conversion from Presto's representation of REAL type to Iceberg's, causing predicating on a REAL type to fail. Queries to reproduce the issue:
```SQL
CREATE TABLE iceberg.u_xinlin.test_real_parquet (_real REAL) WITH (format = 'parquet');
INSERT INTO iceberg.u_xinlin.test_real_parquet VALUES 1.2;
SELECT * FROM iceberg.u_xinlin.test_real_parquet WHERE _real = 1.2;
```
The SELECT query should output 1.2 but it would output no row.

### Commit `Support Files table with nested types`
Iceberg has column stats for nested columns.

### Commit `Collect file stats directly from ORC writers`
When registering a data file with Iceberg, Presto needs to provide file statistics. Previously, Presto would write an ORC file and then reads from it to get the file stats. This commit makes Presto collect file stats directly from ORC writers.

A related issue is that Avro files don't store file stats. So it would be highly-inefficient to re-read a written Avro file and compute the stats. With the new framework, it's possible to implement an Iceberg Avro writer that collects stats on the fly and then uses them when registering a file with Iceberg.

In addition, this commit removes Presto's reference to `iceberg-orc`. This can help us sidestep the ORC version conflict issue when upgrading to `iceberg-0.8.0-incubating`

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -0,0 +1,293 @@
+    public IcebergOrcFileWriter(
+            OrcDataSink orcDataSink,
+            Callable<Void> rollbackAction,
+            Schema icebergSchema,
```
Make this the first or last argument so that the argument ordering matches the parent class (keeps things simpler when reading the code).

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -0,0 +1,293 @@
+        // OrcColumnId(0) is the root column that represents file-level schema
+        for (int i = 1; i < orcColumns.size(); i++) {
+            OrcColumnId orcColumnId = new OrcColumnId(i);
+            if (!excludedColumns.contains(orcColumnId)) {
```
Maybe invert this
```java
if (excludedColumns.contains(orcColumnId)) {
    continue;
}

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -0,0 +1,293 @@
+                    nullCountsBuilder.put(icebergId, fileRowCount - orcColumnStats.getNumberOfValues());
+                }
+                Optional<IcebergMinMax> icebergMinMax = toIcebergMinMax(orcColumnStats, icebergField.type());
+                icebergMinMax.ifPresent(minMax -> {
```
The local variable could be inlined

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -0,0 +1,293 @@
+                OrcType orcColumn = orcColumns.get(orcColumnId);
+                ColumnStatistics orcColumnStats = columnStatistics.get().get(orcColumnId);
+                int icebergId = getIcebergId(orcColumn);
+                Types.NestedField icebergField = requireNonNull(icebergSchema.findField(icebergId), format("Cannot find Iceberg column with ID %d in schema %s", icebergId, icebergSchema));
```
This shouldn't throw NPE. Make this
```java
verify(icebergSchema.findField(icebergId) != null, "Cannot find Iceberg column with ID %s in schema %s", icebergId, icebergSchema);

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -0,0 +1,293 @@
+        Map<Integer, ByteBuffer> lowerBounds = lowerBoundsBuilder.build();
+        Map<Integer, ByteBuffer> upperBounds = upperBoundsBuilder.build();
+        // TODO: Add column size accounting to ORC column writers
+        return new Metrics(fileRowCount,
```
Nit: also put first argument on separate line

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -0,0 +1,293 @@
+        Map<Integer, Long> nullCounts = nullCountsBuilder.build();
+        Map<Integer, ByteBuffer> lowerBounds = lowerBoundsBuilder.build();
+        Map<Integer, ByteBuffer> upperBounds = upperBoundsBuilder.build();
+        // TODO: Add column size accounting to ORC column writers
```
Move this to the null below
```java
new Metrics(
        fileRowCount,
        null, // TODO: Add column size accounting to ORC column writers
```
Or maybe add above like
```java
Map<Integer, Long> columnSizes = null; // TODO: Add column size accounting to ORC column writers

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -0,0 +1,293 @@
+
+    private static void populateExcludedColumns(ColumnMetadata<OrcType> orcColumns, OrcColumnId orcColumnId, boolean shouldBeExcluded, ImmutableSet.Builder<OrcColumnId> excludedColumns)
+    {
+        if (shouldBeExcluded) {
```
Can shorten `shouldBeExcluded` to `exclude`
```java
if (exclude) {

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -0,0 +1,293 @@
+                    populateExcludedColumns(orcColumns, child, shouldBeExcluded, excludedColumns);
+                }
+                return;
+            default:
```
No need for empty default block

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -0,0 +1,293 @@
+
+    private static int getIcebergId(OrcType orcColumn)
+    {
+        String icebergId = requireNonNull(orcColumn.getAttributes().get(ORC_ICEBERG_ID_KEY), format("ORC column %s doesn't have an associated Iceberg ID", orcColumn.toString()));
```
```java
String icebergId = orcColumn.getAttributes().get(ORC_ICEBERG_ID_KEY);
verify(icebergId != null, "ORC column %s doesn't have an associated Iceberg ID", orcColumn);

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -0,0 +1,293 @@
+    {
+        IntegerStatistics integerStatistics = orcColumnStats.getIntegerStatistics();
+        if (integerStatistics != null) {
+            Object min = integerStatistics.getMin();
```
These can be declared as `Long`

lxynov replied:
We can use `Object` so that it can also reference an `Integer`?

electrum replied:
Ah, I missed that the `toIntExact()` below is changing to `Integer`.

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -0,0 +1,293 @@
+                return Optional.empty();
+            }
+            if (icebergType.typeId() == org.apache.iceberg.types.Type.TypeID.INTEGER) {
+                min = toIntExact((Long) min);
```
Cast isn't needed if you change the type

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -0,0 +1,293 @@
+        }
+        DoubleStatistics doubleStatistics = orcColumnStats.getDoubleStatistics();
+        if (doubleStatistics != null) {
+            Object min = doubleStatistics.getMin();
```
`Double`

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergOrcFileWriter.java:
```diff
@@ -0,0 +1,293 @@
+                min = toIntExact((Long) min);
+                max = toIntExact((Long) max);
+            }
+            return Optional.of(IcebergMinMax.builder(icebergType).setMin(min).setMax(max).build());
```
I don't think we need a builder for this. All the usages are for all arguments. So this be
```java
return Optional.of(new IcebergMinMax(icebergType, min, max));
```
Which I find more readable here.

electrum commented on presto-orc/src/main/java/io/prestosql/orc/OrcWriter.java:
```diff
@@ -264,6 +269,7 @@ public void write(Page page)
             }
 
             writeChunk(chunk);
+            fileRowCount += chunkRows;
```
Replace `numberOfRows` in `bufferFileFooter()` with `fileRowCount`

electrum commented on presto-orc/src/main/java/io/prestosql/orc/OrcWriter.java:
```diff
@@ -471,10 +477,11 @@ public void close()
                 .map(ClosedStripe::getStatistics)
                 .map(StripeStatistics::getColumnStatistics)
                 .collect(toList()));
+        fileStats.ifPresent(stats -> fileStatsRetainedBytes = stats.stream().mapToLong(ColumnStatistics::getRetainedSizeInBytes).sum());
```
Assigning a field inside a lambda is not typical and easy to miss, so it'd be better to write like
```java
fileStatsRetainedBytes = fileStats.map(stats -> stats.stream()
        .mapToLong(ColumnStatistics::getRetainedSizeInBytes)
        .sum()).orElse(0L);

lxynov commented:
@electrum Thanks for the review! I've addressed the comments. In addition, a new commit `Fix predicating on REAL type in Iceberg Connector` was added to fix a previous failure in `TestIcebergDistributed`

lxynov commented:
Rebased to master to fix a conflict

electrum commented:
I looks like this test is flaky:
```
[ERROR] testBasic(io.prestosql.plugin.iceberg.TestIcebergOrcMetricsCollection)  Time elapsed: 2.482 s  <<< FAILURE!
java.lang.AssertionError: 
For query: 
 SELECT min(comment) FROM tpch.tiny.orders
not equal
Actual rows (up to 100 of 1 extra rows shown, 1 rows in total):
    [ about the accounts. slyly express accounts wa]
Expected rows (up to 100 of 1 missing rows shown, 1 rows in total):
    [null]

	at org.testng.Assert.fail(Assert.java:94)
	at io.prestosql.testing.QueryAssertions.assertEqualsIgnoreOrder(QueryAssertions.java:237)
	at io.prestosql.testing.QueryAssertions.assertQuery(QueryAssertions.java:203)
	at io.prestosql.testing.QueryAssertions.assertQuery(QueryAssertions.java:106)
	at io.prestosql.testing.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:130)
	at io.prestosql.plugin.iceberg.TestIcebergOrcMetricsCollection.testBasic(TestIcebergOrcMetricsCollection.java:113)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```
I sometimes get the same failure running locally.

lxynov commented:
> this test is flaky:
> ```
> [ERROR] testBasic(io.prestosql.plugin.iceberg.TestIcebergOrcMetricsCollection)  Time elapsed: 2.482 s  <<< FAILURE!
> java.lang.AssertionError: 
> For query: 
> SELECT min(comment) FROM tpch.tiny.orders
> not equal
> Actual rows (up to 100 of 1 extra rows shown, 1 rows in total):
>    [ about the accounts. slyly express accounts wa]
> Expected rows (up to 100 of 1 missing rows shown, 1 rows in total):
>    [null]
> ```
Thanks @electrum for helping debug and find the cause. For people's information, it was because sometimes ORC writer doesn't return string stats due to the limitation on the size of a string stat. (See `maxStringStatisticsLimit` in `OrcWriterOptions`).

@electrum I added one more commit `Fix predicating on DECIMAL type in Iceberg Connector` to fix a test failure in `TestIcebergSmoke::testCreatePartitionedTable` after applying this PR.

electrum commented:
Thanks!

kekwan commented:
@lxynov was there a reason why file stats for ORC timestamp columns were not collected?
